library/common-library/templates/_agent-config.tpl

```diff
@@ -0,0 +1,88 @@
+{{/*
+This helper should return the defaults that the agent should take for all the Helm chart
+*/}}
+{{- define "common.agentConfig.defaults" -}}
+{{- /*if include "common.verboseLog" . }}
+verbose: 1
+{{- end */}}
+{{- if (include "common.nrStaging" .) }}
+staging: true
+{{- end }}
+{{- with include "common.proxy" . }}
+proxy: {{ . | quote }}
+{{- end }}
+{{/* Something with fedramp here */}}
+{{- end -}}
+
+
+
+{{/*
+This helper allow to override the defaults that this common-library adds to all the agents.
+*/}}
+{{- define "common.agentConfig.defaults.override" -}}
+{{/* **********************************************************************************
+***************************************************************************************
+THIS IS AN EXAMPLE OF HOW THE CONFIG DEFAULTS SHOULD GO.
+This should not touch main/master because it has things that could break all the charts
+that use this library.
+nri-kubernetes has been taken as an example with the variables that it has hascoded.
+***************************************************************************************
+********************************************************************************** */}}
+features: "docker_enabled:false"
+
+httpServerEnabled: true
+httpServerPort: 8003
+
+passthroughEnvironment: CLUSTER_NAME
+
+{{- if .Values.enableProcessMetrics | kindIs "bool" }}
+EnableProcessMetrics: {{ .Values.enableProcessMetrics | quote }}
+{{- end }}
+{{- end -}}
+
+
+
+{{/*
+This overrides everything that defaults and user have put in the configuration.
+
+THIS ALSO OVERRIDES USER'S PREFERENCES. USE WITH CARE.
+*/}}
+{{- define "common.agentConfig.override" -}}
+{{/* **********************************************************************************
+***************************************************************************************
+THIS IS AN EXAMPLE OF HOW THE CONFIG DEFAULTS SHOULD GO.
+This should not touch main/master because it has things that could break all the charts
+that use this library.
+nri-kubernetes has been taken as an example with the variables that it has hascoded.
+***************************************************************************************
+********************************************************************************** */}}
+overrideHostRoot: ""
+forwardOnly: true
```

Comment:
I find this a bit scary, there is a lot of logic already involving agent config. 

Aren't enough the helpers for the agent config (proxy, fedramp, verbose, etc) rather then having global config?  Which config are we trying to cover?

Most of agent deployed are used in different ways needing almost no parameters that are not specific to each installation this is why I wonder if it worth it.

Reply:
I answered it [here](https://github.com/newrelic/helm-charts/pull/757#discussion_r832176827) and [here](https://github.com/newrelic/helm-charts/pull/757#discussion_r832171075)

> Is a future proof template that could allow us to use the image from CAOS without having to create new images just tweaking helpers.

> This aims to add flexibility when we failed to add it to the charts so users don't get constraints because of our possible lack of vision of their use cases. [...] We (and the user) can footgun ourselves in many ways like using `verboseLog` at the same time that we change `verbose` in the config. We should document that this is an advance feature, of course.

Reply:
Replacing different images with configuration can make sense. However, wouldn't be a bit obscure for GTS and users having all this logic in helpers?

The point is that I have the feeling that none of the options of the example (and also of the agent) needs to be setted across all charts at the same time with the same value:
```
  features: docker_enabled:false
  forwardOnly: true
  global: global
  global_overriden: test
  httpServerEnabled: true
  httpServerPort: 8003
  isSecureForwardOnly: true
  local: test
  overrideHostRoot: ""
  passthroughEnvironment: CLUSTER_NAME
```

 - Therefore I see this option useful to set the defaults of the agent config in the very same way across multiple charts.

Regarding checking if the config has Verbose or any other config that should be set in a different way we can warn about it

Reply:
You are right. I wrote this to @roobre [here](https://github.com/newrelic/helm-charts/pull/757#discussion_r832176827):

> Maybe this adds noise and you are right. I left it as an example of the way `nri-kubernetes` should use these helpers [...]

I have the feeling that GTS would be more comfortable reading a configMap with the configuration and the `values.yaml` with the agent's configuration than the way we have it right now.

Would you prefer that I create a more finished/polished PR instead of this one and discuss over it?

Reply:
I am still not fully convinced that the use-case is strong enough to add and then maintain all this logic, but maybe I am missing some piece. @roobre what do you think?